scripts/json_parser.py

```diff
@@ -20,44 +22,68 @@ def fix_and_parse_json(json_str: str, try_to_fix_with_gpt: bool = True):
         "criticism": "constructive self-criticism",
         "speak": "thoughts summary to say to user"
     }
-    }
-    """
+}
+"""
+
 
+def fix_and_parse_json(
+    json_str: str,
+    try_to_fix_with_gpt: bool = True
+) -> Union[str, Dict[Any, Any]]:
     try:
         json_str = json_str.replace('\t', '')
         return json.loads(json_str)
-    except Exception as e:
-        # Let's do something manually - sometimes GPT responds with something BEFORE the braces:
-        # "I'm sorry, I don't understand. Please try again."{"text": "I'm sorry, I don't understand. Please try again.", "confidence": 0.0}
-        # So let's try to find the first brace and then parse the rest of the string
+    except json.JSONDecodeError as _:  # noqa: F841
+        json_str = correct_json(json_str)
```

Comment:
`correct_json` tries to parse the message as json in `balance_braces` so if it fails, we never get to the logic below. In my case it failed because the GPT returned a text before valid json, something like "Here's a command that you can use..." Was `correct_json` meant to be included in `try...except` below?

Reply:
```        if balanced_str := balance_braces(json_str):
            return balanced_str```

If the result is None it returns the original json, this is caught by a try/except.

All operations in the correct json are protected with try statements.

Reply:
In `balance_braces` itself you try to parse json and just rethrow an error (which is pointless tbh):
```
```
try:
	json.loads(json_string)
	return json_string
except json.JSONDecodeError as e:
	raise e
```

It is not caught in `correct_json `, it is used in the `except` part without any exception handling

Reply:
Resolved by #747